### ISSUE TITLE
Shared Memory API

### ISSUE THREAD

**Issue by nuta** [nuta/resea#28]
Shared Memory API allows tasks to communicate each others without copying data. Combining the existing IPC APIs, it is a promising approach to copy large data (e.g. file contents) efficiently.

## Interface
In Resea, we can implement shared memory in vm server (I believe we don't need any modifications to the kernel!). The vm should provide the following interface `shm`:

```
# interface.idl

namespace shm {
    rpc create(size: size) -> (shm_id: int);
    rpc map(shm_id: int) -> (vaddr: vaddr);
}
```

## Implementation
Where `create` allocates some physical memory pages for a new shared memory, and `map` maps (by `map_page()`) the specified shared memory (`shm_id`) into an unused virtual address space (by `alloc_virt_pages()`) and returns its address.

We'll need a new data structure to track existing shared memory entries:

```c
struct shm {
    bool in_use;
    int shm_id;
    paddr_t paddr;
    size_t len;
};

#define NUM_SHARED_MEMS_MAX 32
struct shared_mems[NUM_SHARED_MEMS_MAX];
```

## Testing
Add a test in `resea/servers/apps/test`. The test would look like:

```c
void shm_test(void) {
    struct message m;
    m.type = SHM_CREATE_MSG;
    m.size = 256;
    ASSERT_OK(ipc_call(INIT_TASK, &m));

    int shm_id = m.shm_create_reply.shm_id;
    struct message m;
    m.type = SHM_MAP_MSG;
    ASSERT_OK(ipc_call(INIT_TASK, &m));

    // The mapped page should be writable, i.e. this `memset` should not cause a segfault...
    memset((void *) m.shm_map_reply.vaddr, 0x5a, 256);
}
```

## Notes
- This current design does not consider security: anyone can map (or eavesdrop) any shared memory entries. Considering security is a future work.
- This is just a my proposal. Your ideas are welcome!

**Comment by arpitvaghela** [comment]
@nuta  can u elaborate on how `vaddr` is to be allocated, I tried something like this

```c
    m.type = VM_ALLOC_VIRT_PAGES_MSG;
    struct shm_t* shm = shm_stat(shm_id);
    ASSERT_OK(shm != NULL);
    m.vm_alloc_virt_pages.num_pages = shm->len;

    error_t err = ipc_call(INIT_TASK, &m);
    ASSERT_OK(err);
    vaddr_t vaddr = m.vm_alloc_virt_pages_reply.vaddr;
``` 

but this results in a invalid memoy access

**Comment by nuta** [comment]
Could you share your code?

**Comment by arpitvaghela** [comment]
I have partly solved the earlier problem but still the problem persist when i write to the vaddr. 


For test i have implemented a `shm_test server` which creates the `shm`. `shm_test.c` in `test` makes an `ipc call` to get the `shm_id` and `map` the same. [resea/shm_vm](https://github.com/yashrajkakkad/resea/tree/shm_vm)

**Comment by nuta** [comment]
Thanks for sharing! Your implementation looks promising to me.

Regarding. the invalid memory access issue on writes, I guess it's caused by not specifying `MAP_W` in `flag` parameter for `map_page()`.

**Comment by milisarge** [comment]
i wonder, if an another server can access to shared memory without shm_server, for example just trying arbitrary id s

**Comment by nuta** [comment]
Closed by #43
